Update CSP practical guide for consistency (#37227)
mdn/content: files/en-us/web/security/practical_implementation_guides/csp/index.md

## Problem

The main problem this article focuses on is Cross-site scripting ({{Glossary("Cross-site_scripting", "XSS")}}) attacks. These are generally due to a lack of control and awareness of the sources from which site resources are loaded. This problem gets more difficult to manage as sites become larger and more complex and increasingly rely on third-party resources such as JavaScript libraries.

> [!NOTE]
> CSP is one part of a complete strategy for protecting against XSS attacks. There are other factors involved, such as [output encoding](/en-US/docs/Web/Security/Attacks/XSS#output_encoding) and [sanitization](/en-US/docs/Web/Security/Attacks/XSS#sanitization), which are also important.

CSP can also help to fix other problems, which are covered in other articles:

- [Preventing clickjacking](/en-US/docs/Web/Security/Practical_implementation_guides/Clickjacking) by stopping your site being embedded into {{htmlelement("iframe")}} elements. This is done using the CSP [`frame-ancestors`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors) directive.

> - Preventing [manipulator-in-the-middle](/en-US/docs/Glossary/MitM) (MiTM) attacks by upgrading any HTTP connections to HTTPS. This is helped by the CSP [`upgrade-insecure-requests`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors) directive. See [HSTS implementation](/en-US/docs/Web/Security/Practical_implementation_guides/TLS#http_strict_transport_security_implementation) for more details.

Author: updated ;-)

Strict CSPs also:

- Disable the use of unsafe [inline JavaScript](/en-US/docs/Web/HTTP/CSP#inline_javascript), meaning inline [event handler attributes](/en-US/docs/Web/HTML/Attributes#event_handler_attributes) such as `onclick`. This prevents improperly-escaped user inputs from being interpreted by the web browser as JavaScript.
- Disable the use of [risky API calls such as `eval()`](/en-US/docs/Web/HTTP/CSP#eval_and_similar_apis), which is another effect of the `script-src` directive.
- Disable all object embeds via `object-src 'none'`.
- Disable uses of the `<base>` element to set a base URI via `base-uri 'none';`.

Strict CSPs are preferred over [location-based](/en-US/docs/Web/HTTP/CSP#location-based_policies) policies, also called allowlist policies, where you specify which domains scripts can be run from. This is because allowlist policies often end up allowing unsafe domains, which defeats the entire point of having a CSP, and they can get very large and unwieldy, especially if you are trying to permit services that require many third party scripts to function.

> ### Steps for implementing CSP

Author: I disagree; it is not just that — strict CSP is the main thrust of it, but it also covers other things, like alternatives if strict CSP isn't for you, and reporting. I'm going to leave this as-is.

> ### Steps for implementing CSP

Reviewer: Yes, fair enough.

> [!NOTE]
> Before implementing any actual CSP with the `Content-Security-Policy` header, you are advised to first test it out using the [`Content-Security-Policy-Report-Only`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only) HTTP header; see [Report-only CSPs](#report-only_csps) below.

1. Begin by trying out a strict CSP, as outlined above, then start to pinpoint resources that are failing to load as a result of the policy.

> 2. Make sure that external and internal scripts (included via {{htmlelement("script")}} elements) that you want to run have the correct nonce inserted into the [`nonce`](/en-US/docs/Web/HTML/Element/script#nonce) attributes by the server. If you are instead using hashes, external scripts should have the correct hash inserted into [`integrity`](/en-US/docs/Web/HTML/Element/script#integrity) attributes.

Author: OK, I have updated the structure, merging the two bullets into one and adding an extra bullet about choosing nonces or hashes.

> 3. If an allowed script goes on to load other, third-party scripts, those scripts will fail to load because they won't have the required nonce or hash. Mitigate this problem by adding the [`strict-dynamic`](/en-US/docs/Web/HTTP/CSP#the_strict-dynamic_keyword) directive, which gives scripts loaded by the first script the same level of trust without being explicitly given a nonce or hash.

Author: Yup, your suggestion is much better ;-)

Updated.

4. Refactor patterns disallowed by the strict CSP, such as inline event handlers and `eval()`. For example, replace inline event handlers with [`addEventListener()`](/en-US/docs/Web/API/EventTarget/addEventListener) calls inside scripts.
5. Unless sites need the ability to include embeds, their execution should be disabled with `object-src 'none'`.
6. If you are still having trouble with certain items, you can consider an alternative, more permissive policy. For example:
   - Add specific sources as highlighted during testing; for example, `style-src 'self' https://example.com/`.
   - `default-src https:` still provides some protection, disabling unsafe inline/`eval()` and only allowing loading of resources (images, fonts, scripts, etc.) over HTTPS.
   - `unsafe-hashes` allows you to use hashes to enable usage of inline event handlers and `style` attribute values.

>    - `unsafe-eval` allows dynamic evaluation of strings as JavaScript, but beware: it defeats much of the purpose of having a CSP.

Reviewer: Could even pull step 8 out of the list into a subsequent paragraph, to emphasise that this is a departure from the main approach.

> [!WARNING]

> > If at all possible, avoid including unsafe sources inside your CSP. Examples include `unsafe-inline`, `data:` URIs inside `script-src`, `object-src`, or `default-src`, and overly broad sources or form submission targets. Similarly, the use of `script-src 'self'` can be unsafe for sites with JSONP endpoints. These sites should use a `script-src` that includes the path to their JavaScript source folder(s).

Reviewer: ```suggestion
> If at all possible, avoid including unsafe sources inside your CSP. Examples include:
> - `unsafe-inline`
> - `data:` URIs inside `script-src`, `object-src`, or `default-src`
> - overly broad sources or form submission targets
>
> Similarly, the use of `script-src 'self'` can be unsafe for sites with JSONP endpoints. These sites should use a `script-src` that includes the path to their JavaScript source folder(s).
```

Is the JSONP thing only relevant to `self`? I thought it was an issue for any kind of location-based policy. I don't really understand this very well though.

Author: Suggestion applied. I don't know about the JSONP thing either.

Reviewer 2: The JSONP issue is a problem for all sorts host-based lists.

Author: @mozfreddyb looking at https://stackoverflow.com/questions/613962/is-jsonp-safe-to-use#614006, it seems that JSONP is an old hack for achieving cross-domain requests, and we no longer need to do it now we have CORS. This being the case, do we still need to mention it? Do people still do JSONP out in the wild?

If so, what do you think I should say here?

Reviewer 2: Dunno if people use it. Probably just fine to ignore alltogether, I agree.

Author: OK. I've filed https://github.com/mdn/content/pull/37521

If you are unable to use the `Content-Security-Policy` header, pages can instead include a [`<meta http-equiv="Content-Security-Policy" content="…">`](/en-US/docs/Web/HTML/Element/meta#http-equiv) element. This should be the first {{htmlelement("meta")}} element that appears inside the document {{htmlelement("head")}}.

> ## Report-only CSPs

Reviewer: I would have made this an H3.

Author: Yeah, I agree with this. It's not really a standalone topic. Updated.

Before implementing any actual CSP with the `Content-Security-Policy` header, you are advised to first test it out using the [`Content-Security-Policy-Report-Only`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only) HTTP header. This allows you to see if any violations would have occurred with that policy.

Sites should use the [`report-to`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-to) and [`report-uri`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-uri) [reporting directives](/en-US/docs/Glossary/Reporting_directive). These cause the browser to [`POST`](/en-US/docs/Web/HTTP/Methods/POST) JSON reports about CSP violations to endpoints (specified in the {{httpheader("Reporting-Endpoints")}} header in the case of `report-to`). This allows CSP violations to be caught and repaired quickly.

> > [!NOTE] > `report-to` is preferred over the deprecated `report-uri`; however, both are still needed because `report-to` does not yet have full cross-browser support.

Reviewer: I guess the `>` at the start of this is a workaround for the new note format not working when the content starts with a backtick? In this situation Josh uses the old format. You could also do:

```suggestion
> [!NOTE] The `report-to` directive is preferred over the deprecated `report-uri`; however, both are still needed because `report-to` does not yet have full cross-browser support.
```

Author: No, that's just what prettier does to it as a result of the error. I usually fix such instances in the manner you've suggested, but I missed this one. Fixed now!

3. If an allowed script goes on to load other, third-party scripts, those scripts will fail to load because they won't have the required nonce or hash. Mitigate this problem by adding the `strict-dynamic` directive, which gives scripts loaded by the first script the same level of trust without being explicitly given a nonce or hash:

   ```http
   Content-Security-Policy:
     script-src 'nonce-416d1177-4d12-4e3b-b7c9-f6c409789fb8';
     strict-dynamic;
     object-src 'none';
     base-uri 'none';
   ```

   > [!NOTE]
   > This does however weaken the protection offered by strict CSPs. You must make sure that the scripts you are loading are trustworthy.

4. Refactor patterns disallowed by the strict CSP, such as inline event handlers and `eval()`. For example, replace inline event handlers with [`addEventListener()`](/en-US/docs/Web/API/EventTarget/addEventListener) calls inside scripts.
5. If you are still having trouble with certain items, you can consider widening the policy to be more permissive, for example:
   - Adding specific sources as highlighted during testing; for example, `style-src 'self' https://example.com/`.

>    - Adding `default-src https:` still provides some protection, disabling unsafe inline/`eval()` and only allowing loading of resources (images, fonts, scripts, etc.) over HTTPS.

Reviewer: I wonder: if you can deploy a strict CSP at all - that is, if you can add nonces or hashes to your sources - why would you also need to allowlist some sources, or allowlist all HTTPS sources?

Author: Yeah, good point. I have changed this bullet from additions to the policy, to an alternative policy.

>    - Adding `unsafe-hashes` allowsusing hashes to enable usage of inline event handlers and `style` attribute values.

Reviewer: ```suggestion
   - Adding `unsafe-hashes` allows using hashes to enable usage of inline event handlers and `style` attribute values.
```

(although I prefer "allows you to use", "allows using" feels ungrammatical to me)

Author: Yeah, I agree that wording is better; updated.

   - Adding `unsafe-eval` allows dynamic evaluation of strings as JavaScript, but beware: it defeats much of the purpose of having a CSP.

> ## Additional guidance

Reviewer: I'm afraid I don't really have very good suggestions here, but this section feels a bit disorganized. It feels like a list of things about CSP, in any order, some of which are fairly critical security recommendations, some of which are deployment suggestions, some of which are quite niche web compat recommendations.

Some of them connect to and pay out information promised in the previous section, like `report-to` - why isn't this listed in "steps for implementing a CSP"?

Some are already covered in "Solution" but are duplicated here, like `object-src 'none'`.

`data` URL guidance is listed twice, the first time saying "care needs to be taken" (which I don't think is very helpful, what kind of care?) and the second says don't do it.

I think the bit about SVG is super niche and not warranted here.

It might be good to cut out all the duplicated/not needed stuff, put normal deployment things like `report-to` in the previous section, and see if there's a remaining list of "things not to put in your CSP", and if there ius, then add that as a subsection to "Solution".

Author: Yeah, this section was really crappy. I have added a separate section on reporting only CSPs, included some stuff in the "Steps..." section, and deleted/de-duplicated.

Also, it turns out that the Firefox SVG issue has bene fixed (see https://bugzilla.mozilla.org/show_bug.cgi?id=1773976), so I've just removed all that stuff.

{{QuickLinksWithSubpages("/en-US/docs/Web/Security")}}

The [`Content-Security-Policy`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) HTTP header provides fine-grained control over the code that can be loaded on a site, and what it is allowed to do.

## Problem

> 

Author: It is misleading, I agree, but not untrue; one problem is that the coverage was meant to be wider than these HTTP header pages, but for the moment we only have on other guide. ;-(

I've updated the main landing page to improve things here, calling the articles out as summaries, saying they cover "some best practices" and not "the best practices", and changing the subsection to be about "HTTP security fundamentals". I've also changed the description of the CSP article to say it is specifically about mitigating XSS.

## Solution

Implementing a [strict CSP](/en-US/docs/Web/HTTP/CSP#strict_csp) is the best way to prevent XSS vulnerabilities. This uses [nonce-](/en-US/docs/Web/HTTP/CSP#nonces) or [hash-](/en-US/docs/Web/HTTP/CSP#hashes)based fetch directives to ensure that only scripts and/or styles that include the correct nonce or hash will be executed. JavaScript inserted by a hacker will simply not run.

An example is as follows:

```http
Content-Security-Policy:
  script-src 'nonce-416d1177-4d12-4e3b-b7c9-f6c409789fb8';
  object-src 'none';
  base-uri 'none';
```

The nonce should be randomly-generated by your server for every new HTTP response, so would-be attackers cannot guess it.

Strict CSPs also:

> - Disable the use of unsafe [inline JavaScript](/en-US/docs/Web/HTTP/CSP#inline_javascript), meaning inline [event handler attributes](/en-US/docs/Web/HTML/Attributes#event_handler_attributes) such as `onclick`. This prevents improperly-escaped user inputs from being interpreted by the web browser as JavaScript. When a CSP includes a `script-src` directive, _all_ JavaScript must be loaded from external files via {{htmlelement("script")}} elements with `src` attributes.

Reviewer: It's also probably worth mentioning somewhere that if a directive contains nonce or hashes, then AFAIK `unsafe-inline` is ignored (https://web.dev/articles/strict-csp#fallbacks, https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP#inline_javascript).

Which makes me wonder why the observatory seems to complain about unsafe-inline in https://developer.mozilla.org/en-US/observatory/analyze?host=csp-evaluator.withgoogle.com.

> - Disable the use of unsafe [inline JavaScript](/en-US/docs/Web/HTTP/CSP#inline_javascript), meaning inline [event handler attributes](/en-US/docs/Web/HTML/Attributes#event_handler_attributes) such as `onclick`. This prevents improperly-escaped user inputs from being interpreted by the web browser as JavaScript. When a CSP includes a `script-src` directive, _all_ JavaScript must be loaded from external files via {{htmlelement("script")}} elements with `src` attributes.

Author: I've removed the incorrect sentence; I thought I'd done that already, TBH ;-)

I've also added a bullet about `unsafe-inline` being ignored if hashes/nonces are present, later on.

- [Web workers](/en-US/docs/Web/API/Web_Workers_API) (via [document directives](/en-US/docs/Glossary/Document_directive)).
- Navigation / form submission destinations (via [navigation directives](/en-US/docs/Glossary/Navigation_directive)).

### Steps for implementing CSP

> [!NOTE]
> Before implementing any actual CSP with the `Content-Security-Policy` header, you are advised to first test it out using the [`Content-Security-Policy-Report-Only`](/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only) HTTP header. This allows you to see if any violations would have occurred with that policy. This test requires the use of `report-to` (or the deprecated `report-uri`), as explained below.

1. Begin by trying out a strict CSP, as outlined above, then start to pinpoint resources that are failing to load as a result of the policy.
2. Make sure that external and internal scripts (included via {{htmlelement("script")}} elements) that you want to run have the correct nonce or hash inserted into the [`nonce`](/en-US/docs/Web/HTML/Element/script#nonce) or [`integrity`](/en-US/docs/Web/HTML/Element/script#integrity) attributes by the server.

> 3. If an allowed script goes on to load other, third-party scripts, those scripts will fail to load because they won't have the required nonce or hash. Mitigate this problem by adding the `strict-dynamic` directive, which allows the first script to pass its nonce or hash value onto scripts that it loads:

Reviewer: Does "allows the first script to pass its nonce or hash value onto scripts that it loads" imply that there's actual passing on of the value? If so, there isn't: it just makes the trust kind of transitive, so a nonced script is considered trustworthy to dynamically load other scripts, which don't themselves need to be nonced.

   ```http
   Content-Security-Policy:
     script-src 'nonce-416d1177-4d12-4e3b-b7c9-f6c409789fb8';
     strict-dynamic;
     object-src 'none';
     base-uri 'none';
   ```

   > [!NOTE]
   > This does however weaken the protection offered by strict CSPs. You must make sure that the scripts you are loading are trustworthy.

4. Refactor patterns disallowed by the strict CSP, such as inline event handlers and `eval()`. For example, replace inline event handlers with [`addEventListener()`](/en-US/docs/Web/API/EventTarget/addEventListener) calls inside scripts.

> 5. If you are still having trouble with certain items, you can consider widening the policy to be more permissive, adding specific sources as highlighted during testing; for example, `style-src 'self' https://example.com/`. A more permissive policy such as `default-src https:` still provides some protection, disabling unsafe inline/`eval()` and only allow loading of resources (images, fonts, scripts, etc.) over HTTPS, but it is very weak compared to a strict CSP.

Reviewer: Other weakenings, which keep a nonce/hash based approach: `unsafe-eval` makes the policy significantly less secure but still protects you against inline JS. And `unsafe-hashes` if you really can't get rid of inline event handlers: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#unsafe-hashes.

Author: Added info on these
